paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java

```diff
@@ -1274,4 +1308,101 @@ static ConflictCheck noConflictCheck() {
+            Snapshot newSnapshot,
+            boolean useSerializableIsolation,
+            boolean useLock) {
+        try {
+            boolean committed = tryAtomicCommitNewSnapshot(newSnapshotPath, newSnapshot);
+            if (committed) {
+                if (useSerializableIsolation && !useLock) {
+                    // If the useSerializableIsolation is turned on,And we not use lock service,
+                    // we will remove all HINTs,as it may provide incorrect information.
+                    snapshotManager.removeSnapshotEarliestHint();
+                    snapshotManager.removeSnapshotLatestHint();
+                    snapshotManager.removeChangeLogEarliestHint();
+                    snapshotManager.removeChangeLogLatestHint();
+                } else {
+                    snapshotManager.commitLatestHint(newSnapshotId);
+                }
+            }
+            return committed;
+        } catch (Exception e) {
+            throw new CommitStateUnknownException(e.getMessage(), e);
+        }
+    }
+
+    boolean tryAtomicCommitNewSnapshot(Path newSnapshotPath, Snapshot newSnapshot)
+            throws IOException {
+        return fileIO.tryToWriteAtomic(newSnapshotPath, newSnapshot.toJson());
+    }
+
+    void checkBeforeCommit(long newSnapshotId) throws IOException {
+        long latestSnapshotIdBeforeCommit =
+                Optional.ofNullable(snapshotManager.latestSnapshotIdWithOutHint())
+                        .orElse(Snapshot.FIRST_SNAPSHOT_ID - 1);
+        boolean currentCommitIsLatest =
+                Objects.equals(newSnapshotId, latestSnapshotIdBeforeCommit + 1);
+        if (!currentCommitIsLatest) {
+            snapshotManager.removeSnapshotLatestHint();
+            snapshotManager.removeSnapshotEarliestHint();
+            throw new CommitFailedException(
+                    String.format(
+                            "Can't submit snapshotId [%s], because it is expected that snapshotId [%s] should be submitted now.",
+                            newSnapshotId, latestSnapshotIdBeforeCommit + 1));
+        }
+    }
+
+    void checkAfterCommit(long newSnapshotId) throws IOException {
+        String maxSaveSnapshotsNumStr = tableOptions.get(SNAPSHOT_NUM_RETAINED_MAX.key());
+        int maxSaveSnapshotNum = SNAPSHOT_NUM_RETAINED_MAX.defaultValue();
+        if (!StringUtils.isBlank(maxSaveSnapshotsNumStr)) {
+            maxSaveSnapshotNum = Integer.parseInt(maxSaveSnapshotsNumStr);
+        }
+        long latestSnapshotIdAfterCommit =
+                Optional.ofNullable(snapshotManager.latestSnapshotIdWithOutHint())
+                        .orElse(Snapshot.FIRST_SNAPSHOT_ID);
+        long waterMark =
+                latestSnapshotIdAfterCommit > maxSaveSnapshotNum
+                        ? latestSnapshotIdAfterCommit - maxSaveSnapshotNum
+                        : -1;
+        fastFailIfDirtyCommit(newSnapshotId, waterMark, latestSnapshotIdAfterCommit);
```

Comment:
Here we need to add the logic of whether the snapshot belongs to a branch or tag.PR 3787 provides this functionality, I'm not sure if I need to merge it with that PR?

Reply:
For now I'm starting by porting some of the code from 3787 to the current PR.